### PR TITLE
This makes dynamic modules and engines behave themselves much much better

### PR DESCRIPTION
**Pull request by JorisGoosen**
- adds a "reloadData"  command for the engine so that it does that
- Also tries to be more selective about pausing/resuming engines when loading data. Now it is only done for those that actually need the data.
- The rest are just given a signal to reload after the data has changed (such moduleInstallRequest or so)
- Add better event logging from JorisGoosen/jasp-desktop@slowdownHunter
- preferably kill longest idle engine first
- actually log output from R
- Only loading qml when dynamic module is ready for use solves all the crashing it seems
- show a loader while waiting for qml
- Enginewindow added where you can monitor what all the engines are doing (Ctrl+Shift+Alt+E or via prefs-advanced)

- startup better now make sure that shutdown works
- try and keep IPCChannel slaves finding its objects by giving it time. #It doesnt seem to help though and makes everything more complicated :(
-- This is not enough to solve all problems but allows for working around small timing problems
- to further avoid problems the channels are kept separate from the engine and are no long destroyed ever, this way the communication objects aren't lost.
- and anyhow this is only a problem when the max engine count is changed anyway.
- Make sure channel() isnt used when nullptr

- Engines can be killed in enginewindow with middlemouse
- same engine  for installing a module as running it please. This avoids all those "cannot open connection" errors
- make sure forms are destroyed before the rest of qml is so that we arent inundated with errormessages as it exhales its dying breath
- remove useless "initializing" state from Analysis
- fixes dragging issues
- uninstall developer module if path is changed

### CODE REVIEW

File: Desktop/components/JASP/Controls/Form.qml
Comment (JorisGoosen) [inline Desktop/components/JASP/Controls/Form.qml]: @boutinb no right?

Comment (JorisGoosen) [reply Desktop/components/JASP/Controls/Form.qml]: Maybe I do a cleanup round btw, I see some things I might as well remove entirely.
Today or tomorrow.

Comment (boutinb) [reply Desktop/components/JASP/Controls/Form.qml]: Yes this Timer seems to be obsolete.

**Comment by boutinb** [comment]
Your Engines overview is really nice!!
First remark: if my max # engines is 4, and I open more than 4 analyses in different modules, then it kills all engines, and starts again with 1 engine. Why not killing only one engine?

File: Desktop/engine/enginesync.cpp
Comment (boutinb) [inline Desktop/engine/enginesync.cpp:696]: Some extra parenthesis could be nice:
if(engineStopTime != -1 && ((engineStopTime + ENGINE_COOLDOWN) > Utils::currentMillis()) && enginesPossible > 0)

File: Desktop/analysis/analysis.cpp
Comment (boutinb) [inline Desktop/analysis/analysis.cpp:101]: I'm a bit skeptical here: the form is automatically destroyed when its parent is destroyed, so there are some chances that _analysisForm points to a destroyed object.

File: Desktop/analysis/analysis.h
Comment (boutinb) [inline Desktop/analysis/analysis.h]: moduleData() wordt niet meer gebruikt.

File: Desktop/analysis/analysis.cpp
Comment (boutinb) [inline Desktop/analysis/analysis.cpp:130]: I was wondering what exactly is the purpose of this method.
This method is called only by setDynamicModule, which is called only by replaceAnalysesOfDynamicModule, so if I understand well, only when an engine has to process a new module? But if I set a breakpoint in this method, the debugger never stops...

File: Desktop/analysis/analysisform.h
Comment (boutinb) [inline Desktop/analysis/analysisform.h]: _signalValueChangedsBlocked -> _signalValueChangedBlocked

**Review by boutinb — CHANGES_REQUESTED** [review]
Difficult to review the whole code...
I did not spend much time with the shared memory changes, since that will be removed soon.
I have tested it, and it seems to be quite stable

**Comment by JorisGoosen** [comment]
> Your Engines overview is really nice!! First remark: if my max # engines is 4, and I open more than 4 analyses in different modules, then it kills all engines, and starts again with 1 engine. Why not killing only one engine?

Well, there are some timing issues and the analyses were probably all done. Then the engines are idle, it needs 1 but it takes a bit before it is started. Also because I decided to give the shutdown a bit of a cooldown. This to give the boost shared memory file release by the engine some time to complete and avoiding some conflicts through that (aka to hopefully avoid "file not found" etc).

But I can have a look and see if I can tweak it a bit. 

I already noticed another "issue", not really though, where if you install a module it start all allowed engines and then only uses 1 :s 

But that is what JASP always did till now anyway, so its not really a regression I guess.

File: Desktop/analysis/analysis.cpp
Comment (JorisGoosen) [inline Desktop/analysis/analysis.cpp:956]: @boutinb `destroyForm()` always clears `_analysisForm`.

File: Desktop/analysis/analysis.cpp
Comment (JorisGoosen) [reply Desktop/analysis/analysis.cpp:130]: This is necessary because we have multiple datastructures representing the same data :s

The `RibbonModel` has its own copy and `Analysis` links to it, so whenever a module is installed it needs to run this to make sure that it still points to the same place. To be able to do this `Analysis::_codedAnalysisEntry` exists.
If you want the debugger to stop here you should install a developer module. Or change the `Description.qml` of an installed developer module.

File: Desktop/analysis/analysisform.h
Comment (JorisGoosen) [reply Desktop/analysis/analysisform.h]: No this is because you count how many "value changed"s have occurred. This way it is less misleading.
But Im open for better suggestions.

File: Desktop/components/JASP/Controls/Form.qml
Comment (JorisGoosen) [reply Desktop/components/JASP/Controls/Form.qml]: or tomorrow, with the other changes...

File: Desktop/analysis/analysis.cpp
Comment (boutinb) [reply Desktop/analysis/analysis.cpp:956]: Yes, but if the parent of the form is destroyed before, QML will destroy the form without going to this destroyForm method, making _analysisForm a dangling pointer.

Comment (JorisGoosen) [reply Desktop/analysis/analysis.cpp:956]: As we've just talked about, it will be fine

**Comment by boutinb** [comment]
Apparently you forgot to add your changes for enumutilities.h (because it does not compile without).
I have tested it, and I have got these issues:

- In preferences, if you set a wrong path for the developer module, then it crashes when you try to load it
- If you have loaded a developer module, as soon as you change the path in preferences, the developer module is removed. This is quite quick... maybe beter to remove the old one, when the new one is loading.
- if you load a developer module which is not already in the ribbon bar, then it does not come to the ribbon, you need first to click it in the module list. I would set it directly in the ribbon.
- If all engines are set to a module, and you decrease in Preferences the number of engines, it crashes.
- If all engines are set to a module and try to load a developer module of one of the module already used, then you get an error:
Screenshot 2022-03-17 at 13 19 40

**Review by boutinb — APPROVED** [review]

**Comment by boutinb** [comment]
I approve it, but it would be nice to have a loading indicator when a developer module is loaded (and which is not already in the ribbon).
It still crashes when you load a developer module, add and run an analysis from this module, and change the path to the developer module.

**Comment by JorisGoosen** [comment]
> I approve it, but it would be nice to have a loading indicator when a developer module is loaded (and which is not already in the ribbon). 

Well, I agree! So I will fix it.

> It still crashes when you load a developer module, add and run an analysis from this module, and change the path to the developer module.

Oh, that means that the `uninstallJASPModule` function, or whatever its called, is not working perfectly?
Which development module did you load?

**Comment by boutinb** [comment]
> Which development module did you load?

It was crashing with jaspAnova or jaspDistributions module

**Comment by JorisGoosen** [comment]
> > > I approve it, but it would be nice to have a loading indicator when a developer module is loaded (and which is not already in the ribbon).
> > 
> > Well, I agree! So I will fix it.

Well, I still agree but I think I can spend my time more sensibly then on this to be honest. Ill leave it as is.

> > > It still crashes when you load a developer module, add and run an analysis from this module, and change the path to the developer module.
> >
> > Oh, that means that the `uninstallJASPModule` function, or whatever its called, is not working perfectly? Which development module did you load?

> It was crashing with jaspAnova or jaspDistributions module

I cannot reproduce this crash at all. If you really still get this crash I can have a look on my mac on thursday and see whether there is a problem still?

**Comment by JorisGoosen** [comment]
@amirmasoudabdol  did you have a chance to have a look at this PR? I think it would be worthwhile to have another pair of eyes look at this as Ive changed a ton of stuff

**Comment by amirmasoudabdol** [comment]
Not yet, I will, as soon as we are done with the 0.16.2.

File: Desktop/components/JASP/Widgets/RCommanderWindow.qml
Comment (amirmasoudabdol) [inline Desktop/components/JASP/Widgets/RCommanderWindow.qml]: This should be `right`!

File: Desktop/mainwindow.cpp
Comment (amirmasoudabdol) [inline Desktop/mainwindow.cpp:518]: I guess something went wrong with the rebasing, but there should be closing `}` here.

**Comment by amirmasoudabdol** [comment]
While I'm still waiting for the rebase, release, etc., I'm slowing starting to look into this, and I noticed that the RCommander doesn't seem to be working on macOS that I've tested.

File: Desktop/components/JASP/Widgets/EnginesWindow.qml
Comment (amirmasoudabdol) [inline Desktop/components/JASP/Widgets/EnginesWindow.qml]: I think this is going to be problematic. Modern trackpads don't have middle buttons, and often don't have any buttons at all, and that makes it really hard to simulate the middle click. I think we should put an actual button for killing the engine, and  upon click ask the user to make sure.

Comment (JorisGoosen) [reply Desktop/components/JASP/Widgets/EnginesWindow.qml]: Yes that is a very good point! I can add a little `x`-mark in the toprightcorner for instance? Or the red-circle-delete icon from the analyses-view?

The question I won't do because really there is nothing that could go wrong except for waiting slighlty longer for your analysis or whatever to finish. So that would just be useless clicking then.

**Comment by amirmasoudabdol** [comment]
I just noticed that this contains the missing signing identity bug that I introduced, and later fixed! 😬

File: Desktop/components/JASP/Widgets/EnginesWindow.qml
Comment (amirmasoudabdol) [reply Desktop/components/JASP/Widgets/EnginesWindow.qml]: I think the red circle from the analysis should work, and it's also familiar. 

Are you saying that we shouldn't have the terminate functionality at all?

Comment (JorisGoosen) [reply Desktop/components/JASP/Widgets/EnginesWindow.qml]: No Im saying that the terminate functionality is very safe and no work will be lost no matter what.
So that leads me to conclude that having a warning (aka "upon click ask the user to make sure.") is unnecessary.

**Comment by JorisGoosen** [comment]
> I just noticed that this contains the missing signing identity bug that I introduced, and later fixed! 😬

I rebased that into this code? Or where is it?

**Comment by amirmasoudabdol** [comment]
It just needs a rebase to the `stable` at this point, but don't worry about it now, let's have the Qt fixes out first.

**Comment by amirmasoudabdol** [comment]
I think the Engine Overview form needs some work,

- Those circles are massive, and can probably be a bit smaller, and minimal
  - A normal font size should do the job too
- The form probably doesn't need the maximize button, and it can be a Dialog
- The form can be called several times. I think this should not be allowed. 

I noticed that after some times, Engines will be killed, and upon interaction they come back to life. I guess this is something we do, but I've never noticed it! 

<p float="left">



</p>

File: Desktop/components/JASP/Widgets/FileMenu/PrefsAdvanced.qml
Comment (amirmasoudabdol) [inline Desktop/components/JASP/Widgets/FileMenu/PrefsAdvanced.qml]: I'd replace the `#` with "number".

File: Desktop/analysis/analyses.cpp
Comment (amirmasoudabdol) [inline Desktop/analysis/analyses.cpp:55]: I looked a bit around, and I think you might be able to improve the implementation of these `Apply*` methods by using [`std::for_each`](https://en.cppreference.com/w/cpp/algorithm/for_each) or `std::transform`; and if that's doable, then you might be able to get some free parallelism by adding an `ExecutionPolicy` to it. 

I think it might be worth looking into, especially since this is a frequent action.

File: Desktop/analysis/analyses.h
Comment (amirmasoudabdol) [inline Desktop/analysis/analyses.h:187]: This looks like a great candidate for `std::map<size_t, std::unique_ptr<Analysis>>`, or `std::shared_ptr<>` if you really need it. Then, you can just deconstruct them, and don't worry about the `delete` operation, and everything is suddenly great! I mean, of course not, but I feel this is a very good place to start thinking about it. You need a bit of refactoring to make sure that deconstructors are working properly, but then they are safer and more pleasant to work with! 🌻

File: Desktop/analysis/analysisform.h
Comment (amirmasoudabdol) [inline Desktop/analysis/analysisform.h:124]: These `bool` methods should _preferably_ have an `[[nodiscard]]` attribute.

File: Desktop/engine/enginesync.h
Comment (amirmasoudabdol) [inline Desktop/engine/enginesync.h]: I think you probably need one space after `///<` but I'm not sure if it's required. I would argue that it's more readable with the space.

File: Desktop/modules/dynamicmodule.cpp
Comment (amirmasoudabdol) [inline Desktop/modules/dynamicmodule.cpp]: A pair of `()` makes this nicer to read.

File: Desktop/modules/dynamicmodule.cpp
Comment (amirmasoudabdol) [inline Desktop/modules/dynamicmodule.cpp:645]: Then, why bother at all! 😅

File: Desktop/utilities/qutils.cpp
Comment (amirmasoudabdol) [inline Desktop/utilities/qutils.cpp:250]: I guess we want this as an option, if so, I think it can probably go into the `JASP.cmake`.

File: Desktop/widgets/listmodelfactorsform.cpp
Comment (amirmasoudabdol) [inline Desktop/widgets/listmodelfactorsform.cpp]: You can probably use an `std::accumulate` instead.

File: Engine/engine.cpp
Comment (amirmasoudabdol) [inline Engine/engine.cpp:205]: I still need to refactor these into something sensible, but I might not get to them in time. So, I leave this comment in case you want to deal with them in this PR as well! 🙏🏼

**Review by amirmasoudabdol — CHANGES_REQUESTED** [review]
As far as I can tell, this works and I left a few comments/suggestions here and there. If there is anything that needs some attention is the "Engine Overview" form, and the "Engine Kill" behavior that we discussed in one of the other comments before. 

I'm a bit surprised by the fact that we are actively killing the engines, even if there are only a few of them running, but I'm sure there is a good reason for it. However, if we can avoid that, I'd opt for keeping them alive and ready to work, especially if the user has already specified that it wants `X` engines running. 

Also, I'm slightly concerned about the maximum number of engines, ie., 16, but I guess we rarely get to that many anyway!

**Comment by JorisGoosen** [comment]
Thanks for reviewing!


> As far as I can tell, this works and I left a few comments/suggestions here and there. If there is anything that needs some attention is the "Engine Overview" form, and the "Engine Kill" behavior that we discussed in one of the other comments before.

Yeah I know, but it is really only meant for debugging purposes which is why I left it so rough.

> I'm a bit surprised by the fact that we are actively killing the engines, even if there are only a few of them running, but I'm sure there is a good reason for it. However, if we can avoid that, I'd opt for keeping them alive and ready to work, especially if the user has already specified that it wants `X` engines running.

Well, the problem is that R doesn't cleanly unload dlls/dynamic-libs. So the only way to safely reload packages etcetera is by actively killing the engines. In fact, it is the entire purpose of this PR. To make sure the correct stuff is loaded in R by killing it when whatever is loaded isn't needed anymore.

> Also, I'm slightly concerned about the maximum number of engines, ie., 16, but I guess we rarely get to that many anyway!

Well, the default is set at 4 initially so it only grows to 16 if the user changes those settings and a follow-up issue for this is: https://github.com/jasp-stats/jasp-issues/issues/1693

I will start processing your comments from beginning may, but I guess you'll be gone by then.

File: Desktop/components/JASP/Widgets/RCommanderWindow.qml
Comment (JorisGoosen) [reply Desktop/components/JASP/Widgets/RCommanderWindow.qml]: check

File: Desktop/analysis/analyses.h
Comment (JorisGoosen) [reply Desktop/analysis/analyses.h:187]: Well, I like having some control over when things are constructed and deconstructed.
Especially as there are only a few classes where we do this manually anyway. Most of the time we use the whole qobject tree to clean up.

File: Desktop/analysis/analyses.cpp
Comment (JorisGoosen) [reply Desktop/analysis/analyses.cpp:55]: Could be nice but really quite outside of the scope of this pr

File: Desktop/mainwindow.cpp
Comment (JorisGoosen) [reply Desktop/mainwindow.cpp:518]: > I think the Engine Overview form needs some work,
>       * A normal font size should do the job too

A bit strange the font is so huge for you though, are you zooming?

>     * The form probably doesn't need the maximize button, and it can be a Dialog

I guess so, but it also doesnt matter really.

>     * The form can be called several times. I think this should not be allowed.

I agree, but I dont find it important. If I have some time I can see if I can at something to avoid that.
 
> 
> I noticed that after some times, Engines will be killed, and upon interaction they come back to life. I guess this is something we do, but I've never noticed it!

Well, this PR changes a lot of things in the behaviour of engines ;)
This is one of the main goals actually :p

>     * Those circles are massive, and can probably be a bit smaller, and minimal
> the screenshot you show
Hmm the circles are much larger relative to the text for you then for me. 

image
Is what it looks like in default width of the form

File: Desktop/components/JASP/Widgets/FileMenu/PrefsAdvanced.qml
Comment (JorisGoosen) [reply Desktop/components/JASP/Widgets/FileMenu/PrefsAdvanced.qml]: yeah you are right, I dont know why I was trying to keep that as short as possible there is quite some space.

File: Desktop/modules/dynamicmodule.cpp
Comment (JorisGoosen) [reply Desktop/modules/dynamicmodule.cpp]: Sure

File: Desktop/modules/dynamicmodule.cpp
Comment (JorisGoosen) [reply Desktop/modules/dynamicmodule.cpp:645]: Because it might have been triggered while the user was changing it, saved it with a bunch of crap.
Or maybe it just got removed or whatever. You still don't want the entire application to crash and burn then.

File: Desktop/utilities/qutils.cpp
Comment (JorisGoosen) [reply Desktop/utilities/qutils.cpp:250]: Yeah no not really, that was just something I did for debugging but it turned out the problem was somewhere else entirely.
And the code could turn out to be useful so I hid it under that.

File: Desktop/widgets/listmodelfactorsform.cpp
Comment (JorisGoosen) [reply Desktop/widgets/listmodelfactorsform.cpp]: Yes but that is not my code ;)

**Comment by JorisGoosen** [comment]
I've processed the remarks here insofar as I think is useful.

I did however also just fix the permanent installs of modules and still want to do the following:

- Make sure the permanently installed modules are enabled after install. 
- And now they show up at the beginning of the ribbon.  Maybe they should be behind Common? Or even just in the alphabetical location?
- RCommander is broken as well... So I want to fix that too

**Comment by JorisGoosen** [comment]
@boutinb can you verify that the R commander now works for you as well?

These two points:
- Make sure the permanently installed modules are enabled after install.
- And now they show up at the beginning of the ribbon. Maybe they should be behind Common? Or even just in the alphabetical location?

can be done some other time, I think it is time this goes into `development` at least

**Pull request merged**
